dc/s2n-quic-dc/src/stream/server/application.rs

```diff
@@ -0,0 +1,251 @@
+        accept::accept(&self.streams, &self.stats).await
+    }
+
+    #[inline]
+    pub fn acceptor_addr(&self) -> io::Result<SocketAddr> {
+        Ok(self.local_addr)
+    }
+}
+
+/// Default to the SOMAXCONN, similar to rust:
+/// https://github.com/rust-lang/rust/blob/28a58f2fa7f0c46b8fab8237c02471a915924fe5/library/std/src/os/unix/net/listener.rs#L104
+const DEFAULT_BACKLOG: u16 = libc::SOMAXCONN as _;
+
+pub struct Builder {
+    backlog: Option<NonZeroU16>,
+    workers: Option<usize>,
+    acceptor_addr: SocketAddr,
+    span: Option<tracing::Span>,
+    enable_udp: bool,
+    enable_tcp: bool,
+    accept_flavor: accept::Flavor,
+    socket_path: Option<PathBuf>,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            backlog: None,
+            workers: None,
+            // FIXME: Don't default to a fixed port?
+            acceptor_addr: "[::]:4444".parse().unwrap(),
+            span: None,
+            enable_udp: true,
+            enable_tcp: false,
+            accept_flavor: Default::default(),
+            socket_path: None,
+        }
+    }
+}
+
+impl Builder {
+    common_builder_methods!();
+
+    pub fn with_socket_path(mut self, path: &Path) -> Self {
+        self.socket_path = Some(path.to_path_buf());
+        self
+    }
+
+    pub fn build<S: event::Subscriber + Clone>(self, subscriber: S) -> io::Result<Server<S>> {
+        ensure!(
+            self.enable_udp || self.enable_tcp,
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "at least one acceptor type needs to be enabled"
+            ))
+        );
+
+        let concurrency: usize = self.workers.unwrap_or_else(|| {
+            std::thread::available_parallelism()
+                .unwrap_or_else(|_| 1.try_into().unwrap())
+                .into()
+        });
+
+        let backlog: usize = self.backlog.map(NonZeroU16::get).unwrap_or(DEFAULT_BACKLOG) as usize;
+        let (stream_sender, stream_receiver) = mpmc::new::<StreamBuilder<S>>(backlog);
+
+        let env = env::Builder::new(subscriber)
+            .with_threads(concurrency)
+            .with_acceptor(stream_sender.clone());
+
+        let env = env.build()?;
+
+        // TODO is it better to spawn one current_thread runtime per concurrency?
+        let acceptor_rt: runtime::Shared<S> = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .thread_name("uds_worker")
+            .worker_threads(concurrency)
+            .build()?
+            .into();
+
+        let mut span = self.span.unwrap_or_else(tracing::span::Span::current);
+
+        if span.is_none() {
+            span = tracing::debug_span!("server");
+        }
+
+        let (stats_sender, stats_worker, stats) = stats::channel();
+
+        acceptor_rt.spawn(stats_worker.run(env.clock().clone()));
+
+        // Spawn the queue pruner task
```

Comment:
I suspect we don't need this in the application? I think there's not much reason for us to rush on streams that we're pulling from the UDS socket. That should put backpressure indirectly onto the manager process, which will then make the pruner work within the manager if needed. But here we don't actually have a queue I think that's in application space (or we shouldn't).

Reply:
Hm. For some reason I thought the pruner was pruning workers -- I think that's actually a separate mechanism (https://github.com/aws/s2n-quic/blob/73c9278697eba0614ed93dc96d5a8887eeeb1b9a/dc/s2n-quic-dc/src/stream/server/tokio/tcp/manager.rs#L314).

Let's extend the unix datagram packet format by adding the queue_time (e.g., nanoseconds since the unix epoch or something like that). For now, we can just not do anything with it other than deserializing it, but in the future we could use that to prune as we read datagrams from the socket, tracking end-to-end sojourn times from the initial accept() syscall in the manager process. I don't think we need to actively implement that pruning right now though.